java/java.mx.project/src/org/netbeans/modules/java/mx/project/SuiteActionProvider.java

```diff
@@ -139,27 +141,50 @@ public void invokeAction(String action, Lookup context) throws IllegalArgumentEx
                     return;
                 }
                 ListeningDICookie ldic = ListeningDICookie.create(-1);
-                Object obj = ldic.getArgs().get("port"); // NOI18N
                 DebuggerInfo di = DebuggerInfo.create(ListeningDICookie.ID, ldic);
-                DebuggerEngine[] engines = { null };
-                RequestProcessor.getDefault().post(() -> {
-                    DebuggerEngine[] engs = DebuggerManager.getDebuggerManager().startDebugging(di);
-                    engines[0] = engs[0];
+                ASYNC.post(() -> {
+                    DebuggerManager.getDebuggerManager().startDebugging(di);
                 });
                 int port = ldic.getPortNumber();
-                runMx(Bundle.MSG_Unittest(fo.getName()), "--attach", "" + port, "unittest", fo.getName() + testSuffix); // NOI18N
+                running = runBuildAndMx(null, Bundle.MSG_Unittest(fo.getName()), "--attach", "" + port, "unittest", fo.getName() + testSuffix); // NOI18N
                 break;
             default:
                 throw new UnsupportedOperationException(action);
         }
+        if (running != null) {
+            ActionProgress progress = ActionProgress.start(context);
+            running.handle((exitCode, error) -> {
+                progress.finished(exitCode == 0 && error == null);
+                return null;
+            });
+        }
     }
 
-    private boolean runMx(String taskName, String... args) {
+    private CompletionStage<Integer> ensureBuilt(FileObject fo) {
+        if (fo != null) {
+            SuiteSources.Group grp = prj.getSources().findGroup(fo);
+            if (grp != null) {
+                final String name = grp.getDisplayName();
+                return runMx(Bundle.MSG_BuildOnly(prj.getName(), name), "build", "--only", name); // NOI18N
+            }
+        }
+        return runMx(Bundle.MSG_Rebuild(prj.getName()), "build"); // NOI18N
+    }
+
+    private CompletionStage<Integer> runBuildAndMx(FileObject fo, String taskName, String... args) {
+        return ensureBuilt(fo).thenCompose((exitCode) -> {
+            return runMx(taskName, args);
+        });
+    }
+
+    private CompletableFuture<Integer> runMx(String taskName, String... args) {
         final File suiteDir = FileUtil.toFile(prj.getProjectDirectory());
         if (!suiteDir.isDirectory()) {
             Toolkit.getDefaultToolkit().beep();
-            return true;
+            return CompletableFuture.completedFuture(-1);
```

Comment:
The calling method (`invokeAction`) checks for `runMx() != null`; this result could trigger `ActionProgress.start` - OK ?

Reply:
It would trigger start, but also `finish(false)`, right? At least that is consistent...